libbeat/cmd/root.go

```diff
@@ -0,0 +1,63 @@
+package cmd
+
+import (
+	"flag"
+	"os"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// GenRootCmd returns the root command to use for your beat. It takes
+// beat name as paramter, and also run command, which will be called if no args are
+// given (for backwards compatibility)
+func GenRootCmd(name string, beatCreator beat.Creator) *cobra.Command {
+	return GenRootCmdWithRunFlags(name, beatCreator, nil)
+}
+
+// GenRootCmdWithRunFlags returns the root command to use for your beat. It takes
+// beat name as paramter, and also run command, which will be called if no args are
+// given (for backwards compatibility). runFlags parameter must the flagset used by
+// run command
+func GenRootCmdWithRunFlags(name string, beatCreator beat.Creator, runFlags *pflag.FlagSet) *cobra.Command {
+	runCmd := genRunCmd(name, beatCreator, runFlags)
+
+	rootCmd := &cobra.Command{
+		Use: name,
+		Run: runCmd.Run,
+	}
+
+	// Persistent flags, common accross all subcommands
+	rootCmd.PersistentFlags().AddGoFlag(flag.CommandLine.Lookup("E"))
+	rootCmd.PersistentFlags().AddGoFlag(flag.CommandLine.Lookup("c"))
+	rootCmd.PersistentFlags().AddGoFlag(flag.CommandLine.Lookup("d"))
+	rootCmd.PersistentFlags().AddGoFlag(flag.CommandLine.Lookup("path.config"))
+	rootCmd.PersistentFlags().AddGoFlag(flag.CommandLine.Lookup("path.data"))
+	rootCmd.PersistentFlags().AddGoFlag(flag.CommandLine.Lookup("path.logs"))
+	rootCmd.PersistentFlags().AddGoFlag(flag.CommandLine.Lookup("path.home"))
+	rootCmd.PersistentFlags().AddGoFlag(flag.CommandLine.Lookup("strict.perms"))
+
+	// Inherit root flags from run command
+	// TODO deprecate when root command no longer executes run (7.0)
+	rootCmd.Flags().AddFlagSet(runCmd.Flags())
+
+	// Register subcommands common to all beats
+	rootCmd.AddCommand(runCmd)
+	rootCmd.AddCommand(genVersionCmd(name))
+
+	return rootCmd
+}
+
+func init() {
```

Comment:
I would recommend to put the `init` on top, at least that is what we did so far.